pkg/controllers/user/serviceaccount/serviceaccount.go

```diff
@@ -0,0 +1,323 @@
+				// no project is associated with namespace so don't do anything
+				logrus.Infof("no project is associated with namespace %v", namespaceName)
+				return nil
+			}
+
+			split := strings.Split(annotation, ":")
+
+			if len(split) != 2 {
+				return fmt.Errorf("could not parse project id annotation: %v", annotation)
+			}
+
+			clusterName, projectId := split[0], split[1]
+
+			projects, err := m.projectLister.List("", labels.Everything())
+
+			var project *v3.Project
+
+			for _, candidate := range projects {
+				if candidate.Name == projectId {
+					project = candidate
+					break
+				}
+			}
+
+			if err != nil {
+				return fmt.Errorf("error getting projects: %v", err)
+			}
+
+			podSecurityPolicyTemplateId := project.Spec.PodSecurityPolicyTemplateName
+
+			if podSecurityPolicyTemplateId == "" {
+				// check cluster
+				cluster, err := m.clusterLister.Get("", clusterName)
+
+				if err != nil {
+					return fmt.Errorf("error getting clusters: %v", err)
+				}
+
+				if cluster == nil {
+					return fmt.Errorf("could not find cluster %v", clusterName)
+				}
+
+				podSecurityPolicyTemplateId = cluster.Spec.DefaultPodSecurityPolicyTemplateName
+
+				if podSecurityPolicyTemplateId == "" {
+					logrus.Infof("No PSPTs found for project %v and cluster %v", projectId, clusterName)
+					return nil
+				}
+			}
+
+			var policy *v1beta12.PodSecurityPolicy
+
+			for _, candidate := range policies {
+				if candidate.Annotations[podSecurityAnnotation] == podSecurityPolicyTemplateId {
+					policy = candidate
+				}
+			}
+
+			if policy == nil {
+				template, err := m.templateLister.Get("", podSecurityPolicyTemplateId)
+
+				if err != nil {
+					return fmt.Errorf("error getting pod security policy templates: %v", err)
+				}
+
+				if template == nil {
+					return fmt.Errorf("no template exists with id %v", podSecurityPolicyTemplateId)
+				}
+
+				template.Kind = "PodSecurityPolicy"
+				template.APIVersion = apiVersion
+				template.ResourceVersion = ""
+
+				objectMeta := template.ObjectMeta
+				objectMeta.Annotations[podSecurityAnnotation] = template.Name
+
+				policy, err = m.policies.Create(&v1beta12.PodSecurityPolicy{
+					TypeMeta:   template.TypeMeta,
+					ObjectMeta: objectMeta,
+					Spec:       template.Spec,
+				})
+
+				if err != nil {
+					return fmt.Errorf("error creating pod security policy: %v", err)
+				}
+
+				logrus.Info("successfully created a pod security policy!")
+			}
+
+			roles, err := m.roleLister.List("", labels.Everything())
+
+			if err != nil {
+				return fmt.Errorf("error getting roles: %v", err)
+			}
+
+			var role *rbac.Role
+
+			for _, candidate := range roles {
+				if candidate.Annotations[podSecurityAnnotation] == policy.Name {
+					role = candidate
+				}
+			}
+
+			if role == nil {
+				// Create role
+				newRole := &rbac.Role{
+					ObjectMeta: metav1.ObjectMeta{
+						Annotations: map[string]string{},
+						Name:        fmt.Sprintf("%v-role", policy.Name),
```

Comment:
If you dont need a consistent name for lookups (seems like you shouldnt if you are looking up via a ~annotation~ label), then use
```
GeneratedName: "role-",
```
instead of `Name: ...`